cupy/_core/core.pyx

```diff
@@ -1838,7 +1892,21 @@ cdef class _ndarray_base:
             else:
                 a_gpu = self
             a_cpu = out
-        else:
+
+        if a_cpu is None:
+            # we don't check is_ump_supported() etc here because it'd be
+            # done later
+            if _is_ump_enabled:
+                try:
+                    # return self to use the same memory and avoid copy
+                    a_cpu = numpy.asarray(self, order=order)
+                except TypeError:
```

Comment:
IIRC I only check if the user has set the env var to enable zero copy, but I skip (defer) the check for hw support until the buffer protocol is tried. If the hw does not support zero copy, buffer creation would fail, and NumPy ends up raising this error.

Reply:
Hmm, I tried to raise `RuntimeError` in `__getbuffer__` but not getting TypeError here. Am I missing something?

```py
# dummy_buf.pyx
# To compile: cythonize -b -3 dummy_buf.pyx

import numpy as np

cdef class Dummy:
    def __getbuffer__(self, Py_buffer *buffer, int flags):
        print('__getbuffer__ called')
        raise RuntimeError
    def __releasebuffer__(self, Py_buffer *buffer):
        print('__releasebuffer__ called')
        raise RuntimeError

def test():
    d = Dummy()
    print(d)
    a = np.asarray(d)
    print(type(a), a.dtype)
```

```sh
$ python -c 'import dummy_buf; dummy_buf.test()'
<dummy_buf.Dummy object at 0x7fe3e718ac00>
__getbuffer__ called
<class 'numpy.ndarray'> object
```

Reply:
@kmaehashi sorry I posted this in the wrong thread. I think I missed the second half of the story (sorry it's been a while since I worked on this feature...). `TypeError` is raised in `__array__`:
https://github.com/cupy/cupy/blob/678add80cdd82e89f2b751f1722035c53997ad03/cupy/_core/core.pyx#L1489-L1493
which can be easily reproduced by adding a dummy `def __array__(self): raise TypeError` to your snippet above. We do not allow implicitly converting CuPy arrays to NumPy (object) arrays.

Reply:
Ah, got it, thanks!